tedana/tests/test_utils.py

```diff
@@ -80,73 +80,130 @@ def test_make_adaptive_mask():
     mask_file = pjoin(datadir, "mask.nii.gz")
     data = io.load_data(fnames, n_echos=len(tes))[0]
 
+    # Add in simulated values
+    base_val = np.mean(data[:, 0, :])  # mean value of first echo
+    idx = 5457
+    # Three good echoes (3)
+    data[idx, :, :] = np.array([base_val, base_val - 1, base_val - 2])[:, None]
+    # Dropout: good bad good (3)
+    # Decay: good good bad (2)
+    data[idx + 1, :, :] = np.array([base_val, 1, base_val])[:, None]
+    # Dropout: good bad bad (1)
+    # Decay: good good bad (2)
+    data[idx + 2, :, :] = np.array([base_val, 1, 1])[:, None]
+    # Dropout: good good good (3)
+    # Decay: good bad bad (1)
+    data[idx + 3, :, :] = np.array([base_val, base_val, base_val])[:, None]
+    # Dropout: bad bad bad (0)
+    # Decay: good good good (3)
+    data[idx + 4, :, :] = np.array([1, 0.9, 0.8])[:, None]
+    # Base: good good bad (2)
+    # Dropout: bad bad bad (0)
+    # Decay: good good good (3)
+    data[idx + 5, :, :] = np.array([1, 0.9, -1])[:, None]
+
     # Just dropout method
-    mask, masksum = utils.make_adaptive_mask(
+    mask, adaptive_mask = utils.make_adaptive_mask(
         data,
         mask=mask_file,
         threshold=1,
         methods=["dropout"],
     )
 
-    assert mask.shape == masksum.shape == (64350,)
-    assert np.allclose(mask, (masksum >= 1).astype(bool))
-    assert mask.sum() == 49376
-    vals, counts = np.unique(masksum, return_counts=True)
+    assert mask.shape == adaptive_mask.shape == (64350,)
+    assert np.allclose(mask, (adaptive_mask >= 1).astype(bool))
+    assert adaptive_mask[idx] == 3
+    assert adaptive_mask[idx + 1] == 3
+    assert adaptive_mask[idx + 2] == 1
+    assert adaptive_mask[idx + 3] == 3
+    assert adaptive_mask[idx + 4] == 0
+    assert adaptive_mask[idx + 5] == 0
+    assert mask.sum() == 49374
+    vals, counts = np.unique(adaptive_mask, return_counts=True)
     assert np.allclose(vals, np.array([0, 1, 2, 3]))
-    assert np.allclose(counts, np.array([14974, 3682, 5128, 40566]))
+    assert np.allclose(counts, np.array([14976, 1817, 4427, 43130]))
```

Comment:
I was trying to get a better handle with what was happening within this PR. In `make_adaptive_mask` I added `older_mask =(np.abs(echo_means) > lthrs).sum(axis=-1)` right after `dropout_adaptive_mask` was calculated and compared results in my debugger. Calls like `((older_mask==2) * (dropout_adaptive_mask==2)).sum()` show voxels that used to be 1 in the adaptive mask and are now 2.
Voxels that were 0 are still 0. (This interacts with other masking steps, but seems matched at this point in the code)
Of the 3590 voxels that were 1, 1061 are now 2, and 913 are now 3.
Of the voxels 5276 voxels that were 2, 1900 are now 3.
As expected, none of the voxels that had a higher values are now lower.

That means this change will substantively expand the number of voxels used in ICA and will balance out some of the drop caused by raising the threshold from masking. The one thing that concerns me is it turns how this test data had 913 voxels where the first and second echos were both below threshold, but the third was above threshold. That doesn't seem great, but the proposed enhancement would give this voxels a 0 in the adaptive mask, which would mean they'd be even be excluded from the optimally combined data. There's nothing to change here, but this is a discussion that will be relevant if a future enhancement is considered.

Reply:
I've opened #1083 about this. Can you follow up in that issue with your ratio idea?